### ISSUE TITLE
v8.1.0834 breaks LeaderF in terminal vim

### ISSUE THREAD

**Issue by bennyyip** [vim/vim#3874]
Problem:
Since patch patch 8.1.0834(8.1.0833 is fine), Leaderf window freezes when it shows up, cannot input anything.
This bug only appears in terminal vim, GVim is ok.

Steps to reproduce:
1. Place https://github.com/Yggdroot/LeaderF in ~/.vim/pack/bug/start/LeaderF
2. Open vim in terminal
3. `:Leaderf file`

Environment:
- Linux
- Alacritty/Konsole/Tmux
- Build flags: 
```
CFLAGS='-g -O2' STRIP=true ./configure --prefix=$HOME/.local --localstatedir=$HOME/.local/var/lib/vim --mandir=$HOME/.local/share/man --enable-acl --disable-gui --without-x --disable-fontset
--disable-xsmp --enable-pythoninterp --enable-perlinterp=dynamic --enable-cscope --enable-multibyte --with-features=huge --enable-terminal --enable-fail-if-missing
```

**Comment by brammool** [comment]
> Problem:
> Since patch patch 8.1.0834(8.1.0833 is fine), Leaderf window freezes when it shows up, cannot input anything.
> 
> Steps to reproduce:
> 1. Place https://github.com/Yggdroot/LeaderF in ~/.vim/pack/bug/start/LeaderF
> 2. Open vim in terminal
> 3. `:Leaderf file`
> 
> Environment:
> - Linux
> - Build with: 
> ```
> CFLAGS='-g -O2' STRIP=true ./configure --prefix=$HOME/.local --localstatedir=$HOME/.local/var/lib/vim --mandir=$HOME/.local/share/man --enable-acl --disable-gui --without-x --disable-fontset
> --disable-xsmp --enable-pythoninterp --enable-perlinterp=dynamic --enable-cscope --enable-multibyte --with-features=huge --enable-terminal --enable-fail-if-missing
> ```

Can you narrow down where/why this happens?
Does typing CTRL-C do anything?

I thought this CL didn't change anything for the terminal, only for the
GUI.  I cannot guess why it makes a difference for this plugin.

If the plugin is using a job and/or channel, you can get more info by
doing this before typing ":Leaderf":

	call ch_logfile('debuglog', 'w')

And see what's in the debuglog file.

-- 
Vim is like Emacs without all the typing.  (John "Johann" Spetz)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ichizok** [comment]
@brammool 

Cause:
The following "wait_time" comparison, changed in CLI between v8.1.0833 (in mch_inchar) and v8.1.0834, affected the behavior on TUI.

https://github.com/vim/vim/commit/e40b9d47bf8f8f716d3ef5a95c8ecbbdc0a501f9#diff-b68adb4fa34020d8d7f0ab40a2704335L416 (in mch_inchar)

```c
	    if (wait_time < 0)
		    {
			if (wtime >= 0)
			    /* no character available within "wtime" */
			    return 0;
```

https://github.com/vim/vim/commit/e40b9d47bf8f8f716d3ef5a95c8ecbbdc0a501f9#diff-56b46d1a5b12250be3f04930aef53484R316 (in inchar_loop)

```c
	    if (wait_time <= 0)
		    {
			if (wtime >= 0)
			    // no character available within "wtime"
			    return 0;
```

**Comment by ichizok** [comment]
Simple example:

test.vim
```vim
function GetChar()
  let c = 0
  while c == 0
    let c = getchar(0)
    sleep 10m
  endwhile
  return c
endfunction
```

```
vim --clean -S test.vim
:echo GetChar()
```

Vim doesn't react to user input except for Ctrl-C.
That is, now getchar(0) always returns 0.

**Comment by brammool** [comment]
> Simple example:
> 
> test.vim
> ```vim
> function GetChar()
>   let c = 0
>   while c == 0
>     let c = getchar(0)
>     sleep 10m
>   endwhile
> endfunction
> ```
> 
> ```
> vim --clean -S test.vim
> :call GetChar()
> ```
> 
> Vim doesn't react to user input except for Ctrl-C.
> That is, now getchar(0) always returns 0.

Thanks, now I understand what's happening.

I changed the check for zero, because in the GUI it did wait for zero
time, it hangs.  But with the change it never calls the wait function.
We can make the GUI wait for a very short time.


-- 
WOMAN:   Well, 'ow did you become king then?
ARTHUR:  The Lady of the Lake, [angels sing] her arm clad in the purest
         shimmering samite, held aloft Excalibur from the bosom of the water
         signifying by Divine Providence that I, Arthur, was to carry
         Excalibur.  [singing stops] That is why I am your king!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by skywind3000** [comment]
@bennyyip , is it fixed in 8.1.840 ??
https://github.com/vim/vim/commit/12dfc9eef14fe74c46145aa9e6cba9666f1bcd40

**Comment by bennyyip** [comment]
@skywind3000 fixed
